scintillometry/psrio/psrfits_io.py

```diff
@@ -0,0 +1,383 @@
+from astropy.io import fits
+import numpy as np
+
+
+__all__ = ["PsrfitsHearderHDU", "SubintHDU"]
+
+
+class HDUBase:
+    """This is the HDUBase class that defines some generic API functions
+    """
+    _header_defaults = {}
+    def verify(self):
+        raise NotImplementError
+
+    def set_header_card(self, key, value=None, comment=None):
+        if value is None and key in self._header_defaults.keys():
+            self.header.set(key, self._header_defaults['key']['value'],
+                            self._header_defaults['key']['comment'])
+        else:
+            self.header.set(key, value, comment)
+
+class PsrfitsHearderHDU(fits.PrimaryHDU, HDUBase):
+    """HeaderHDU class provides the translator function between baseband-style
+    file object and the PSRFITS main header HDU.
+
+    Parameter
+    ---------
+    header_hdu : `Pdat.header_hdu`
+    """
+    _properties = ('start_time', 'observatory', 'frequency', 'ra', 'dec',
+                   'shape', 'sample_rate')
+
+    _header_defaults = main_header
+
+    def __init__(self, header_hdu=None):
+        if header_hdu is None:
+            super(PsrfitsHearderHDU, self).__init__()
+            self._init_empty()
+        else:
+            super(PsrfitsHearderHDU, self).__init__(header=header_hdu.header,
+                                                    data=header_hdu.data)
+            self.verify()
+
+    def verify(self):
+        if 'SIMPLE' not in list(self.header.keys()):
+            raise ValueError("The input HDU is not a fits headers HDU.")
+        if self.header['FITSTYPE'] != "PSRFITS":
+            raise ValueError("The input fits header is not a PSRFITS type.")
+
+    def _init_empty(self):
+        for k, v in self._header_defaults.items():
+            self.header.set(k, v['value'], v['comment'])
+        self.header.add_comment('FITS (Flexible Image Transport System) format'
+                                'is defined in Astronomy and Astrophysics, '
+                                'volume 376, page 359; bibcode:'
+                                ' 2001A&A...376..359H', after='EXTEND')
+
+    @property
+    def start_time(self):
+        # TODO add location
+        MJD_d_int = np.longdouble(self.header['STT_IMJD']) * u.day
+        MJD_s_int = np.longdouble(self.header['STT_SMJD']) * u.s
+        MJD_s_frac = np.longdouble(self.header['STT_OFFS']) * u.s
+        #NOTE I am assuming PSRFITS's time scale is UTC
+        return Time(MJD_d_int, MJD_s_int + MJD_s_frac, format='mjd',
+                    scale='utc', precision=9)
+
+    @start_time.setter
+    def start_time(self, time):
+        # TODO add location
+        if not isinstance(time, Time):
+            raise ValueError("Input time should be an `astropy.time.Time`"
+                             " object.")
+        if not time.isscalar:
+            raise ValueError("Input time should be a scalar")
+        else:
+            self.header['STT_IMJD'] = int(divmod(time.mjd, 1)[0])
```

Comment:
The `divmod` should not be needed.

Note that this can suffer from rounding errors; see `baseband/data/header.py` (`start_time.setter`) for what should be a bit more secure implementation (other parts of `baseband` may have similar problems...)

Reply:
I updated it. testing it will be in the writing pull request.